### PR TITLE
config support 2nd try

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

File: services/src/util/config.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    pub list_limit: u32,
+}
+
+impl ConfigElement for ProjectService {
+    const KEY: &'static str = "project_service";
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GdalSource {
+    pub raster_data_root_path: String, // TODO: Path?
```
Comment (jdroenner) [inline services/src/util/config.rs:108]: yes paths everywhere

File: services/src/util/config.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            .map(File::from)
+            .collect();
+
+        settings.merge(files).unwrap();
+
+        settings
+    });
+}
+
+pub fn get_config<'a, T>(key: &str) -> Result<T>
```
Comment (jdroenner) [inline services/src/util/config.rs:40]: will this de-serialize the config on each call?

**Review by jdroenner — COMMENTED** [review]
looks good so far. How does config handle get() calls? Will it always de-serialize?

File: services/src/util/config.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/util/config.rs:40]: good point. Yes it deserializes on each call. Fortunately, it doesn't read the file each time ;)

It also only deserializes lazily once the config key is accessed. So we might end up with the problem, where a faulty configuration is not detected until the value is actually used for the first time.

Maybe we should just bundle all the configs in a Config struct and deserialize into it from the config file once at startup. Then we at least know that all required values are present and valid. I am not sure if we even need the config crate in this case. It makes it at least easy to merge multiple config files like Settings-default.toml and Settings.toml

File: services/src/util/config.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/util/config.rs:108]: I could use `PathBuf` as `Path` is not owned. However it doesn't seems to check if it's really a valid path in the file system. So makes no differences to using string

File: services/src/util/config.rs (hunk 2 above)
Comment (jdroenner) [reply services/src/util/config.rs:40]: We could deserialize into a typed map/list. Then we can keep the get<T> thing. I also think that a validation of the config when it is loaded is important.

File: services/src/util/config.rs (hunk 1 above)
Comment (jdroenner) [reply services/src/util/config.rs:108]: i think PathBuf is fine since String is also an owned type.

File: services/src/util/config.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/util/config.rs:40]: but what ist the type of the value of the map? `Map<String, ?>`. Do we put all config structs in an enum?

Comment (jdroenner) [reply services/src/util/config.rs:40]: we will have to use a crate like https://docs.rs/typemap/0.3.3/typemap/ or build something similar

File: services/src/util/config.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/util/config.rs]: e44b19d69a1fa7d2452c8255a59f491b6b6353c3

File: services/src/util/config.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/util/config.rs:40]: I created a task for this https://dbs-projects.mathematik.uni-marburg.de/T1227

**Pull request merged**